File: bench/cpu/pointer_chasing.zig
const std = @import("std");

const N = 100_000;

fn printResult(name: []const u8, total_ns: u64, iters: u64) !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {d:>8} ns {d:>12}\n", .{ name, total_ns / iters, iters });
}

const Node = struct {
    value: i32,
    next: ?*Node,
};

pub fn main() !void {
    const alloc = std.heap.page_allocator;
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {s:>12} {s:>12}\n", .{ "Benchmark", "Time", "Iterations" });
    try w.print("{s}\n", .{"-" ** 66});

    // Linked list (random pointer chasing)
    {
        const nodes = try alloc.alloc(Node, N);
        defer alloc.free(nodes);

        var indices: [N]usize = undefined;
        for (&indices, 0..) |*x, i| x.* = i;
        var prng = std.Random.DefaultPrng.init(42);
        prng.random().shuffle(usize, &indices);

        for (0..N - 1) |i| {
            nodes[indices[i]] = .{ .value = @intCast(indices[i]), .next = &nodes[indices[i + 1]] };
        }
        nodes[indices[N - 1]] = .{ .value = @intCast(indices[N - 1]), .next = null };

        const head: *Node = &nodes[indices[0]];

        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i32 = 0;
            var p: ?*Node = head;
            while (p) |node| {
                sum +%= node.value;
                p = node.next;
            }
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_LinkedList", total, iters);
    }

    // Array traversal
    {
        const arr = try alloc.alloc(i32, N);
        defer alloc.free(arr);
        for (arr, 0..) |*x, i| x.* = @intCast(i);

        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i32 = 0;
            for (arr) |x| sum +%= x;
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_ArrayTraversal", total, iters);
    }
}

File: bench/cpu/pointer_chasing.cpp
#include <benchmark/benchmark.h>

#include <algorithm>
#include <numeric>
#include <random>
#include <vector>

struct Node {
  int value;
  Node* next;
};

static void BM_LinkedList(benchmark::State& state) {
  constexpr int N = 100'000;
  std::vector<Node> nodes(N);

  std::vector<int> indices(N);
  std::iota(indices.begin(), indices.end(), 0);
  std::mt19937 rng(42);
  std::shuffle(indices.begin(), indices.end(), rng);

  for (int i = 0; i < N - 1; ++i) {
    nodes[indices[i]].value = indices[i];
    nodes[indices[i]].next = &nodes[indices[i + 1]];
  }
  nodes[indices[N - 1]].value = indices[N - 1];
  nodes[indices[N - 1]].next = nullptr;

  Node* head = &nodes[indices[0]];

  for (auto _ : state) {
    int sum = 0;
    Node* p = head;
    while (p) {
      sum += p->value;
      p = p->next;
    }
    benchmark::DoNotOptimize(sum);
  }
}

static void BM_ArrayTraversal(benchmark::State& state) {
  constexpr int N = 100'000;
  std::vector<int> arr(N);
  std::iota(arr.begin(), arr.end(), 0);

  for (auto _ : state) {
    int sum = 0;
    for (auto i = 0u; i < arr.size(); ++i) {
      sum += arr[i];
    }
    benchmark::DoNotOptimize(sum);
  }
}

BENCHMARK(BM_LinkedList);
BENCHMARK(BM_ArrayTraversal);

BENCHMARK_MAIN();
